flytekit/remote/backfill.py

```diff
@@ -0,0 +1,83 @@
+import logging
+import typing
+from datetime import datetime, timedelta
+
+from croniter import croniter
+
+from flytekit import LaunchPlan
+from flytekit.core.workflow import ImperativeWorkflow, WorkflowBase
+from flytekit.remote.entities import FlyteLaunchPlan
+
+
+def create_backfill_workflow(
+    start_date: datetime,
+    end_date: datetime,
+    for_lp: typing.Union[LaunchPlan, FlyteLaunchPlan],
+    parallel: bool = False,
+    per_node_timeout: timedelta = None,
+    per_node_retries: int = 0,
+) -> typing.Tuple[WorkflowBase, datetime, datetime]:
+    """
+    Generates a new imperative workflow for the launchplan that can be used to backfill the given launchplan.
+    This can only be used to generate  backfilling workflow only for schedulable launchplans
+
+    the Backfill plan is generated as (start_date - exclusive, end_date inclusive)
+
+    :param start_date: datetime generate a backfill starting at this datetime (exclusive)
+    :param end_date: datetime generate a backfill ending at this datetime (inclusive)
+    :param for_lp: typing.Union[LaunchPlan, FlyteLaunchPlan] the backfill is generatd for this launchplan
+    :param parallel: if the backfill should be run in parallel. False (default) will run each bacfill sequentially
+    :param per_node_timeout: timedelta Timeout to use per node
+    :param per_node_retries: int Retries to user per node
+    :return: WorkflowBase, datetime datetime -> New generated workflow, datetime for first instance of backfill, datetime for last instance of backfill
+    """
+    if not for_lp:
+        raise ValueError("Launch plan is required!")
+
+    if start_date >= end_date:
+        raise ValueError(
+            f"for a backfill start date should be earlier than end date. Received {start_date} -> {end_date}"
+        )
+
+    schedule = for_lp.entity_metadata.schedule if isinstance(for_lp, FlyteLaunchPlan) else for_lp.schedule
+
+    if schedule is None:
+        raise ValueError("Backfill can only be created for scheduled launch plans")
+
+    if schedule.cron_schedule is not None:
+        cron_schedule = schedule.cron_schedule
+    else:
+        raise NotImplementedError("Currently backfilling only supports cron schedules.")
+
+    logging.info(f"Generating backfill from {start_date} -> {end_date}. Parallel?[{parallel}]")
+    wf = ImperativeWorkflow(name=f"backfill-{for_lp.name}")
+    date_iter = croniter(cron_schedule.schedule, start_time=start_date, ret_type=datetime)
+    prev_node = None
+    actual_start = None
+    actual_end = None
+    while True:
+        next_start_date = date_iter.get_next()
```

Comment:
I'm not familiar with `croniter`. Can you an example of when a call to `get_next` is _not_ return a valid date?

Reply:
hmm how is that possible for a valid croniter? It is a date iterator. we will always go into the future right?

Reply:
nvm, I re-read the code and it's ok.